java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java

```diff
@@ -0,0 +1,148 @@
+@SqaleConstantRemediation("15min")
+public class PublicStaticMutableMembersCheck extends SubscriptionBaseVisitor {
+
+  private static final Set<String> MUTABLE_TYPES = ImmutableSet.of(
+    "java.awt.Point",
+    "java.util.Date",
+    "java.util.Collection",
+    "java.util.Map"
+  );
+
+  // java.util and apache commons
+  private static final MethodInvocationMatcherCollection UNMODIFIABLE_METHOD_CALLS = MethodInvocationMatcherCollection.create()
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.anyType()).name(NameCriteria.startsWith("unmodifiable")).withNoParameterConstraint())
+      // apache commons 3.X
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections.map.UnmodifiableMap")).name("decorate").withNoParameterConstraint())
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections.set.UnmodifiableSet")).name("decorate").withNoParameterConstraint())
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections.list.UnmodifiableList")).name("decorate").withNoParameterConstraint())
+      // apache commons 4.X
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections4.map.UnmodifiableMap")).name("decorate").withNoParameterConstraint())
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections4.set.UnmodifiableSet")).name("decorate").withNoParameterConstraint())
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections4.list.UnmodifiableList")).name("decorate").withNoParameterConstraint());
+  private static final MethodMatcher ARRAYS_AS_LIST = MethodMatcher.create()
+    .typeDefinition("java.util.Arrays").name("asList").withNoParameterConstraint();
+
+  private static final Set<String> ACCEPTED_TYPES = ImmutableSet.of(
+    "com.google.common.collect.ImmutableMap",
+    "com.google.common.collect.ImmutableCollection"
+  );
+  
+  private static final Set<String> ACCEPTED_NEW_TYPES = ImmutableSet.of(
+    "org.apache.commons.collections4.list.UnmodifiableList"
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.ENUM);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    List<Tree> members = ((ClassTree) tree).members();
+    for (Tree member : members) {
+      if (member.is(Tree.Kind.VARIABLE)) {
+        VariableTree variableTree = (VariableTree) member;
+        Symbol symbol = variableTree.symbol();
+        if (symbol != null && isPublicStatic(symbol) && isForbiddenType(symbol.type()) && isNotImmutable(variableTree.initializer())) {
+          addIssue(member, "Make this member \"protected\".");
+        }
+      }
+    }
+  }
+
+  private static boolean isNotImmutable(@Nullable ExpressionTree initializer) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "isNotImmutable" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)